File: python/path-in-zigzag-labelled-binary-tree.py
class Solution:
    def get_tree_level(label: int) -> int:
        level = 0
        while 0 < label:
            level += 1
            label //= 2
        return level

    def get_inversed(value: int, level: int) -> int:
        level_min = 2 ** (level - 1)
        level_max = level_min * 2 - 1
        position = (level_max - value) + level_min
        return position

    def get_label_or_position(value: int, level: int) -> int:
        if level % 2 == 0:
            value = Solution.get_inversed(value, level)
        return value

    def pathInZigZagTree(self, label: int) -> List[int]:
        level = Solution.get_tree_level(label)
        path = []
        while 0 < level:
            path.append(label)
            parent_position = Solution.get_label_or_position(label, level) // 2
            level -= 1
            label = Solution.get_label_or_position(parent_position, level)
        path.reverse()
        return path

File: go/path-in-zigzag-labelled-binary-tree.go
func GetTreeLevel(label int) int {
	level := 0
	for 0 < label {
		level += 1
		label /= 2
	}
	return level
}

func GetInverse(value int, level int) int {
	if level <= 1 {
		return 1
	}
	levelMin := 1 << (level - 1)
	levelMax := (levelMin << 1) - 1
	position := (levelMax - value) + levelMin
	return position
}

func GetLabelOrPosition(value int, level int) int {
	if (level & 1) == 0 {
		value = GetInverse(value, level)
	}
	return value
}

func Reverse(path []int) {
	n := len(path) / 2
	for i := 0; i < n; i++ {
		j := len(path) - i - 1
		aux := path[i]
		path[i] = path[j]
		path[j] = aux
	}
}

func pathInZigZagTree(label int) []int {
	level := GetTreeLevel(label)
	path := make([]int, 0)
	for 0 < level {
		path = append(path, label)
		parentPosition := GetLabelOrPosition(label, level) / 2
		level -= 1
		label = GetLabelOrPosition(parentPosition, level)
	}
	Reverse(path)
	return path
}
